src/frontend/src/expr/function_call.rs

```diff
@@ -115,6 +115,30 @@ impl FunctionCall {
                 }
                 align_types(inputs.iter_mut())
             }
+            ExprType::ConcatWs => {
+                if inputs.len() < 2 {
+                    return Err(ErrorCode::BindError(
+                        "ConcatWs function must contain at least 2 arguments".into(),
+                    )
+                    .into());
+                }
+
+                if inputs[0].return_type() != DataType::Varchar {
+                    return Err(ErrorCode::BindError(
+                        "ConcatWs function must have text as first argument".into(),
+                    )
+                    .into());
+                }
+
+                // subsequent inputs can be any type, they are cast into varchars with
+                // explicit_cast.
+                inputs = inputs
+                    .into_iter()
+                    .map(|input| input.cast_explicit(DataType::Varchar))
+                    .collect::<Result<Vec<_>>>()?;
```

Comment:
```
                inputs = inputs
                    .into_iter()
                    .enumerate()
                    .map(|(i, input)| match i == 0 {
                        // 0-th arg must be string
                        true => input.cast_implicit(DataType::Varchar),
                        // subsequent can be any type
                        false => input.cast_explicit(DataType::Varchar),
                    })
                    .collect::<Result<Vec<_>>>()?;
```
This would also allow the e2e test cases that use literal `null` as separator.

Reply:
Thank you! I was wondering how to patch that.